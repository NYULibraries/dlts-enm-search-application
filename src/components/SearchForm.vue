<template>
    <section class="hero is-primary enm-searchbox-hero">
        <div class="hero-body">
            <div class="container is-fluid">
                <form
                    class="enm-searchform"
                    @submit.prevent="submitSearchForm">
                    <div class="enm-searchbox">
                        <div class="field columns level">
                            <div class="column level-item">
                                <p class="control has-icons-left">
                                    <label
                                        class="enm-visually-hidden"
                                        for="enm-searchinput">Search inside all books</label>
                                    <input
                                        id="enm-searchinput"
                                        v-model="queryUI"
                                        class="input is-large"
                                        type="text"
                                        placeholder="Search inside all books">
                                    <span class="icon is-small is-left">
                                        <i class="fa fa-search"></i>
                                    </span>
                                </p>
                            </div>

                            <div class="column is-narrow">

                                <template v-for="queryFieldUI in queryFieldsUI">
                                    <input
                                        :name="queryFieldUI.name + 'Chx'"
                                        :id="queryFieldUI.name + 'Chx'"
                                        :key="queryFieldUI.name"
                                        :value="queryFieldUI.value"
                                        v-model="selectedQueryFields"
                                        type="checkbox"
                                        class="is-medium is-checkbox">
                                    <label
                                        :for="queryFieldUI.name + 'Chx'"
                                        :key="queryFieldUI.label"
                                        class="">
                                        {{ queryFieldUI.label }}
                                    </label>
                                </template>

                            </div>
                        </div>
                    </div>
                </form>
            </div>
        </div>
    </section>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';

export default {
    name: 'SearchForm',
    props : {
        queryFieldsUI : {
            type     : Array,
            required : true,
            default  : null,
        },
    },
    data() {
        return {
            queryUI : null,

            selectedQueryFields: this.queryFieldsUI.map(
                ( queryField ) => { return queryField.value; }
            ),
        };
    },
    computed : {
        ...mapGetters(
            [
                'query',
                'queryFields',
                'selectedTopicFacetItems',
            ]
        ),
    },
    methods: {
        ...mapActions(
            [
                'addSelectedTopicFacetItems',
                'removeSelectedTopicFacetItems',
                'setQuery',
                'setQueryFields',
            ]
        ),
        submitSearchForm() {
            if ( this.selectedQueryFields.length === 0 ) {
                alert( 'Please check one or more boxes: ' +
                       this.queryFieldsUI.map(
                           ( e ) => { return e.label; }
                       ).join( ', ' )
                );

                return;
            }

            this.setQuery( this.queryUI );
            this.$emit( 'submit', this.query, this.selectedQueryFields );
        },
    },
};
</script>

<style>
</style>
