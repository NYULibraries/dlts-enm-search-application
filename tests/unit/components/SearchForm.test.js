import SearchForm from '@/components/SearchForm';
import { createLocalVueWithVuex, queryFieldsUI, queryFieldsUILabels } from '../test-utils';

import mergeWith from 'lodash.mergeWith';
import { shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';

const QUERY                      = 'something';
const QUERY_FIELDS_FULL_TEXT     = 'pageText';
const QUERY_FIELDS_TOPIC_NAMES   = 'topicNames';
const QUERY_FIELDS_ALL           = Object.freeze( [ QUERY_FIELDS_FULL_TEXT, QUERY_FIELDS_TOPIC_NAMES ] );

function createWrapper( storeOverrides, mountingOverrides ) {
    const localVue = createLocalVueWithVuex();
    const defaultStoreOptions = {
        actions : {
            setQuery                     : () => {},
            setQueryFields               : () => {},
        },
        getters : {
            query                   : () => QUERY,
            queryFields             : () => QUERY_FIELDS_ALL,
        },
    };

    // Allow for empty objects and arrays in overrides to replace default values
    function customizer( objValue, srcValue ) {
        return srcValue || objValue;
    }

    const store = new Vuex.Store(
        mergeWith( defaultStoreOptions, storeOverrides, customizer )
    );

    const defaultMountingOptions = {
        propsData : {
            queryFieldsUI : queryFieldsUI(),
        },
        localVue,
        store,
    };

    const mergedMountingOptions = mergeWith(
        defaultMountingOptions, mountingOverrides, customizer
    );

    return shallowMount( SearchForm, mergedMountingOptions );
    // return shallowMount( SearchForm, merge( defaultMountingOptions, mountingOverrides ) );
}

describe( 'SearchForm', () => {
    describe( 'renders correctly when initialized', () => {
        test( 'with no queryFields', () => {
            const storeOverrides = {
                getters : {
                    queryFields : () => [],
                },
            };
            const mountingOverrides = {
                propsData : {
                    queryFieldsUI : [],
                },
            };
            const wrapper = createWrapper( storeOverrides, mountingOverrides );

            expect( wrapper.element ).toMatchSnapshot();
        } );

        test( 'with queryFields', () => {
            expect( createWrapper().element ).toMatchSnapshot();
        } );
    } );

    test( 'calls window.alert with correct arguments ' +
               'when search submitted with no query fields selected', () => {

        const storeOverrides = {
            getters : {
                query : () => '',
            },
        };

        const wrapper = createWrapper( storeOverrides );

        const spyAlert = jest.spyOn( window, 'alert' ).mockImplementation( () => {
            // noop
        } );

        queryFieldsUI().map( queryFieldUI => queryFieldUI.name )
            .forEach( ( queryField ) => {
                wrapper.find( `input[ id = "${ queryField }Chx" ]` ).setChecked( false );
            } );

        wrapper.find( 'form' ).trigger( 'submit' );

        expect( spyAlert ).toHaveBeenCalledWith(
            'Please check one or more boxes: ' + queryFieldsUILabels().join( ', ' )
        );

        spyAlert.mockRestore();
    } );
} );
