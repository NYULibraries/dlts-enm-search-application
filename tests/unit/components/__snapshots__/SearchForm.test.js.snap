// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly when initialized with no queryFields 1`] = `
<section
  class="hero is-primary enm-searchbox-hero"
>
  <div
    class="hero-body"
  >
    <div
      class="container is-fluid"
    >
      <form
        class="enm-searchform"
      >
        <div
          class="enm-searchbox"
        >
          <div
            class="field columns level"
          >
            <div
              class="column level-item"
            >
              <p
                class="control has-icons-left"
              >
                <label
                  class="enm-visually-hidden"
                  for="enm-searchinput"
                >
                  
                                    Search inside all books
                                
                </label>
                 
                <input
                  class="input is-large"
                  id="enm-searchinput"
                  placeholder="Search inside all books"
                  type="text"
                />
                 
                <span
                  class="icon is-small is-left"
                >
                  <i
                    class="fa fa-search"
                  />
                </span>
              </p>
            </div>
             
            <div
              class="column is-narrow"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
</section>
`;

exports[`SearchForm renders correctly when initialized with queryFields 1`] = `
<section
  class="hero is-primary enm-searchbox-hero"
>
  <div
    class="hero-body"
  >
    <div
      class="container is-fluid"
    >
      <form
        class="enm-searchform"
      >
        <div
          class="enm-searchbox"
        >
          <div
            class="field columns level"
          >
            <div
              class="column level-item"
            >
              <p
                class="control has-icons-left"
              >
                <label
                  class="enm-visually-hidden"
                  for="enm-searchinput"
                >
                  
                                    Search inside all books
                                
                </label>
                 
                <input
                  class="input is-large"
                  id="enm-searchinput"
                  placeholder="Search inside all books"
                  type="text"
                />
                 
                <span
                  class="icon is-small is-left"
                >
                  <i
                    class="fa fa-search"
                  />
                </span>
              </p>
            </div>
             
            <div
              class="column is-narrow"
            >
              <input
                class="is-medium is-checkbox"
                id="fulltextChx"
                name="fulltextChx"
                type="checkbox"
                value="pageText"
              />
               
              <label
                for="fulltextChx"
              >
                
                                    Full Text
                                
              </label>
              <input
                class="is-medium is-checkbox"
                id="topicsChx"
                name="topicsChx"
                type="checkbox"
                value="topicNames"
              />
               
              <label
                for="topicsChx"
              >
                
                                    Topics
                                
              </label>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</section>
`;

exports[`SearchForm when search is submitted renders correctly 1`] = `
<section
  class="hero is-primary enm-searchbox-hero"
>
  <div
    class="hero-body"
  >
    <div
      class="container is-fluid"
    >
      <form
        class="enm-searchform"
      >
        <div
          class="enm-searchbox"
        >
          <div
            class="field columns level"
          >
            <div
              class="column level-item"
            >
              <p
                class="control has-icons-left"
              >
                <label
                  class="enm-visually-hidden"
                  for="enm-searchinput"
                >
                  
                                    Search inside all books
                                
                </label>
                 
                <input
                  class="input is-large"
                  id="enm-searchinput"
                  placeholder="Search inside all books"
                  type="text"
                />
                 
                <span
                  class="icon is-small is-left"
                >
                  <i
                    class="fa fa-search"
                  />
                </span>
              </p>
            </div>
             
            <div
              class="column is-narrow"
            >
              <input
                class="is-medium is-checkbox"
                id="fulltextChx"
                name="fulltextChx"
                type="checkbox"
                value="pageText"
              />
               
              <label
                for="fulltextChx"
              >
                
                                    Full Text
                                
              </label>
              <input
                class="is-medium is-checkbox"
                id="topicsChx"
                name="topicsChx"
                type="checkbox"
                value="topicNames"
              />
               
              <label
                for="topicsChx"
              >
                
                                    Topics
                                
              </label>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</section>
`;
